sandbox/cgroup/monitor.go

```diff
@@ -0,0 +1,189 @@
+			folderList = append(folderList, fullPath)
+		} else {
+			iw.removePath(fullPath)
+		}
+	}
+	if len(folderList) == 0 {
+		newWatch.channel <- newWatch.name
+	} else {
+		newWatch.folders = folderList
+		iw.groupList = append(iw.groupList, newWatch)
+	}
+}
+
+func (iw *inotifyWatcher) removePath(fullPath string) {
+	iw.pathList[fullPath]--
+	if iw.pathList[fullPath] == 0 {
+		iw.wd.RemoveWatch(fullPath)
+		delete(iw.pathList, fullPath)
+	}
+}
+
+// processDeletedPath checks if the received "path corresponds to the passed
+// CGroup, removing it from the list of folders being watched in that CGroup if
+// needed. It returns TRUE if there remain folders to be monitored in that CGroup,
+// or FALSE if all the folders of that CGroup have been deleted.
+func (iw *inotifyWatcher) processDeletedPath(watch *groupToWatch, deletedPath string) bool {
+	for i, fullPath := range watch.folders {
+		if fullPath == deletedPath {
+			// if the folder name is in the list of folders to monitor, decrement the
+			// parent's usage counter, and remove it from the list of folders to watch
+			// in this CGroup (by not adding it to tmpFolders)
+			iw.removePath(fullPath)
+			watch.folders = append(watch.folders[:i], watch.folders[i+1:]...)
+		}
+	}
+	if len(watch.folders) == 0 {
+		// if all the files/folders of this CGroup have been deleted, notify the
+		// client that it is done.
+		watch.channel <- watch.name
+		return false
+	}
+	return true
+}
+
+func (iw *inotifyWatcher) watcherMainLoop() {
+	for {
+		select {
+		case event := <-iw.wd.Event:
+			if event.Mask&inotify.InDelete == 0 {
+				continue
+			}
+			var newGroupList []*groupToWatch
+			for _, watch := range iw.groupList {
+				if iw.processDeletedPath(watch, event.Name) {
+					newGroupList = append(newGroupList, watch)
+				}
+			}
+			iw.groupList = newGroupList
+		case newWatch := <-iw.addWatchChan:
+			iw.addWatch(newWatch)
+		}
+	}
+}
+
+// MonitorDelete monitors the specified paths for deletion.
+// Once all of them have been deleted, it pushes the specified name through the channel.
+func (iw *inotifyWatcher) monitorDelete(folders []string, name string, channel chan<- string) error {
+	iw.doOnce.Do(func() {
+		wd, err := inotify.NewWatcher()
+		if err == nil {
+			iw.wd = wd
+			go iw.watcherMainLoop()
+		}
+	})
+
+	if iw.wd == nil {
+		return errors.New("cannot initialise Inotify.")
+	}
+	iw.addWatchChan <- &groupToWatch{
+		name:    name,
+		folders: folders,
+		channel: channel,
+	}
+	return nil
+}
+
+// MonitorSnapEnded is the method to call to monitor the running instances of an specific Snap.
+// It receives the name of the snap to monitor (for example, "firefox" or "steam")
+// and a channel. The caller can wait on the channel, and when all the instances of
+// the specific snap have ended, the name of the snap will be sent through the channel.
+// This allows to use the same channel to monitor several snaps
+func MonitorSnapEnded(snapName string, channel chan string) error {
```

Comment:
If needed, we can add an "AbortSnapMonitor" function...

Reply:
Alternatively `MonitorSnapEnded()` could become `func MonitorSnapEnded(snapName string) (<-chan string, error) {` and this side controls the channel lifetime? The way the API is used later in https://github.com/snapcore/snapd/pull/12355/files?diff=unified&w=0#diff-f94ae2bf132e8fc6ff519dee7285334af9c3490625817c15b6610f72e1d18b47R652 indicates this is fine?

Reply:
It is possible, yes, but... does that really change anything? I mean... If the main program closes the received channel, it will still panic, isn't it?

Reply:
(unless, of course, that we are talking about "conventions"...)

Reply:
Yeah, I think it's fine, was wondering a bit alout but I think we should go forward with things as they are for now.

Reply:
What about adding a function that just sends the string over the channel, and includes a recover in case the code panics because the channel is closed?

Reply:
When I wrote this, I considered the possibility of the caller sharing the same channel between monitored cgroups (that's why the cgroup name is sent over the channel).